### PR TITLE
Handle the error case when a container reattaches to the same network

### LINKED ISSUE #632: Running docker container has no ipAddress
* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

### Expected behavior
Connecting a container to a network it's already connect should do nothing.
### Actual behavior
It does something weird and removes the IPAddress of the container in that network.
* restarting the container solves the issue.
### Steps to reproduce the behavior
```shell
sudip@Linux-Mint:~$ docker network create --driver overlay --attachable docker
sudip@Linux-Mint:~$ docker run -d --name docker alpine sleep 5m
sudip@Linux-Mint:~$ docker network connect docker docker
sudip@Linux-Mint:~$ docker container inspect docker | grep IPAddress
            "SecondaryIPAddresses": null,
            "IPAddress": "",
                    "IPAddress": "10.0.2.8",
sudip@Linux-Mint:~$ docker network connect docker docker
Error response from daemon: endpoint with name docker already exists in network docker
sudip@Linux-Mint:~$ docker container inspect docker | grep IPAddress
            "SecondaryIPAddresses": null,
            "IPAddress": "",
                    "IPAddress": "",
sudip@Linux-Mint:~$ docker restart docker
docker
sudip@Linux-Mint:~$ docker container inspect docker | grep IPAddress
            "SecondaryIPAddresses": null,
            "IPAddress": "",
                    "IPAddress": "10.0.2.8",

```

**Output of `docker version`:**

```
Client:
 Version:           18.09.1
 API version:       1.39
 Go version:        go1.10.6
 Git commit:        4c52b90
 Built:             Wed Jan  9 19:35:31 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.1
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       4c52b90
  Built:            Wed Jan  9 19:02:44 2019
  OS/Arch:          linux/amd64
  Experimental:     false

```

**Output of `docker info`:**

```
Containers: 11
 Running: 6
 Paused: 0
 Stopped: 5
Images: 323
Server Version: 18.09.1
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: active
 NodeID: o1ueyyf01njcs1pi14ud1yqqf
 Is Manager: true
 ClusterID: adngjwcxr5w5rky9lcu1r7lzu
 Managers: 1
 Nodes: 1
 Default Address Pool: 10.0.0.0/8  
 SubnetSize: 24
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 192.168.1.139
 Manager Addresses:
  192.168.1.139:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 96ec2177ae841256168fcf76954f7177af9446eb
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-20-generic
Operating System: Linux Mint 19
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.42GiB
Name: Linux-Mint
ID: CXCB:6FON:JOCL:CREP:ROLV:5PLY:HU7U:WVD2:SQ3Z:B6EU:MOTX:HCL3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

### PR DESCRIPTION
**Pull request by arkodg**
fixes https://github.com/docker/for-linux/issues/632

Signed-off-by: Arko Dasgupta <arko.dasgupta@docker.com>

**- What I did**

Made sure the Networks context is manipulated in the scenario when a container attempts to attach to an a network it is already connected to

**- How I did it**
When attaching to a network in `findAndAttachNetwork` make sure you return with an error when that network is already part of `container.NetworkSettings.Networks`

**- How to verify it**
Followed the same steps in the issue and made sure it is resolved

**- Description for the changelog**
Fix an issue where connecting a container to a network it's already connected to removes its IP address.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by arkodg** [comment]
@mavenugo PTAL

**Comment by thaJeztah** [comment]
Wondering; should we actually error, or treat this as a no-op (i.e., make network connect be idempotent)?

File: daemon/container_operations.go
Comment (thaJeztah) [inline daemon/container_operations.go]: If we do not want to handle this silently (note that if we do, we should still check if the provided custom IP-addresses (if any)) are the same as the existing ones, otherwise produce an error), I think this should be a `types.BadRequestErrorf`, not a "forbidden" (HTTP 403)

Actually, a `Conflict` (HTTP 409) might be more appropriate https://github.com/moby/moby/blob/c19b9b1885ec6c9a9fa91211111691f994aef5b7/errdefs/defs.go#L13-L17


Also, instead of handling this situation here, could we delegate this it to `daemon.clusterProvider.AttachNetwork` ? It feels like this is something that that function should handle https://github.com/moby/moby/blob/d912e9460e287238f879de791595193543ab2202/daemon/cluster/networks.go#L181-L182

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: daemon/container_operations.go
Comment (arkodg) [reply daemon/container_operations.go]: thanks for the suggestion, I'd rather return an rrror and not silently return, a `409` seems more apt

**Comment by arkodg** [comment]
Derek add label: rebuild/*

**Comment by arkodg** [comment]
Derek add label:rebuild/*

**Comment by cpuguy83** [comment]
Wow, this is weird.
We are already erroring out here, I'm wondering why network settings is being modified in that case.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
The issue seems to be that the way the code is structured we are modifying the container object before things have even happened.
I think this is a good validation and happy to get it in... sadly it seems like the core logic deeper in the stack is flawed and difficult to work with.

**Comment by andrewhsu** [comment]
Looked at this at the maintainers meeting with @arkodg and @cpuguy83 and this PR looks like it will help catch an error case but need bigger refactor to address the private function behaviours of `moby/daemon/container_operations.go` which will be tracked in a new issue.

**Comment by cpuguy83** [comment]
Ping need another reviewer.

**Comment by arkodg** [comment]
@thaJeztah PTAL

**Review by thaJeztah — APPROVED** [review]
whoops, my bad, thought I left mine already

LGTM

**Comment by tiborvass** [comment]
CI passed 5 days ago, but merging this broke master because of https://github.com/moby/moby/pull/39332

**Pull request merged**
